File: 05-Algorithms/code/python/seb_alg/z_algorithm.py
"""Mirrors `typescript/src/zAlgorithm.ts` exactly."""

from __future__ import annotations

from typing import List


def z_algorithm(s: str) -> List[int]:
    """Z-array: z[i] is the longest common prefix length between s and
    s[i:]. z[0] is conventionally len(s). O(|s|).
    """
    n = len(s)
    z = [0] * n
    if n == 0:
        return z
    z[0] = n
    l, r = 0, 0
    for i in range(1, n):
        if i < r:
            z[i] = min(r - i, z[i - l])
        while i + z[i] < n and s[z[i]] == s[i + z[i]]:
            z[i] += 1
        if i + z[i] > r:
            l, r = i, i + z[i]
    return z

File: 05-Algorithms/code/typescript/src/zAlgorithm.ts
/**
 * Computes the Z-array of `s`: `z[i]` is the length of the longest common
 * prefix between `s` and `s.slice(i)` (`z[0]` is conventionally the full
 * string length, though it carries no useful information for matching).
 * O(|s|).
 */
export function zAlgorithm(s: string): number[] {
  const n = s.length;
  const z = new Array(n).fill(0);
  if (n === 0) return z;
  z[0] = n;
  let l = 0;
  let r = 0;
  for (let i = 1; i < n; i++) {
    if (i < r) {
      z[i] = Math.min(r - i, z[i - l]);
    }
    while (i + z[i] < n && s[z[i]] === s[i + z[i]]) z[i]++;
    if (i + z[i] > r) {
      l = i;
      r = i + z[i];
    }
  }
  return z;
}
